torch_datasets.py

```diff
@@ -0,0 +1,194 @@
+from allennlp.data.token_indexers import SingleIdTokenIndexer, TokenIndexer
+
+logger = logging.getLogger(__name__)
+
+class Dataset(TorchDataset, Registrable):
+    def __init__(self):
+        self.vocab: Vocabulary = None
+
+    def text_to_instance(self, *inputs) -> Instance:
+
+        raise NotImplementedError
+
+    def __getitem__(self) -> Instance:
+
+        raise NotImplementedError
+
+    def index_with(self, vocab: Vocabulary):
+        self.vocab = vocab
+
+
+"""
+Here we have two SNLI readers in both of the different styles.
+They are only slightly different.
+"""
+
+
+class SnliDataset(Dataset):
+    def __init__(
+        self, file_path: str, token_indexers: Dict[str, TokenIndexer] = None, lazy: bool = False
+    ) -> None:
+
+        super().__init__()
+
+        self._tokenizer = lambda x: [Token(t) for t in x.split(" ")]
+        self._token_indexers = token_indexers or {"tokens": SingleIdTokenIndexer()}
+
+        file_path = cached_path(file_path)
+        self.examples = []
+
+        for line in open(file_path, "r"):
+            example = json.loads(line)
+            if example["gold_label"] == "-":
+                # These were cases where the annotators disagreed; we'll just skip them.  It's
+                # like 800 out of 500k examples in the training data.
+                continue
+            self.examples.append(example)
+
+    def __len__(self):
+        return len(self.examples)
+
+    def __getitem__(self, idx) -> Instance:
+        example = self.examples[idx]
+        instance = self.text_to_instance(
+            example["sentence1"], example["sentence2"], example["gold_label"]
+        )
+
+        # This is not ideal, we don't want a user to have to worry about this
+        # but at the same time, it's expensive and it would be nice if it could happen here.
+        # It's possible we could have this in the super class.
+        if self.vocab is not None:
+            instance.index_fields(self.vocab)
+        return instance
+
+    @overrides
+    def text_to_instance(self, premise: str, hypothesis: str, label: str = None) -> Instance:
+
+        fields: Dict[str, Field] = {}
+        premise_tokens = self._tokenizer(premise)
+        hypothesis_tokens = self._tokenizer(hypothesis)
+        fields["premise"] = TextField(premise_tokens, self._token_indexers)
+        fields["hypothesis"] = TextField(hypothesis_tokens, self._token_indexers)
+        if label:
+            fields["label"] = LabelField(label)
+
+        metadata = {
+            "premise_tokens": [x.text for x in premise_tokens],
+            "hypothesis_tokens": [x.text for x in hypothesis_tokens],
+        }
+        fields["metadata"] = MetadataField(metadata)
+        return Instance(fields)
+
+
+class BatchInstanceSampler(BatchSampler):
+    
+    def __init__(self, data, batch_size: int, sorting_keys: List[Tuple[str, str]] = None, padding_noise: float = 0.1):
```

Comment:
Also better here if there's something that separates the configuration from the data, so we can easily use the same configuration on multiple datasets (e.g., train and dev).